### ISSUE TITLE
OpenAPI deepObject serialization for query parameters

### ISSUE THREAD

**Issue by FlorianLudwig** [tiangolo/fastapi#3163]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Implement a Pull Request for a confirmed bug.

### Description

Allow usage of `deepObject` [serialization from OpenAPI 3](https://swagger.io/docs/specification/serialization/).

### What it could look like

```Python
from fastapi import FastAPI

app = FastAPI()

class UserQuery(BaseModel):
    role: str
    firstName: str

@app.get("/")
def read_root(id: Optional[UserQuery]):
    return {"Hello": "World"}
```

Query:
```
/users?id[role]=admin&id[firstName]=Alex
```

### Related

 * https://github.com/tiangolo/fastapi/issues/245#issuecomment-762729917
   * Links to a good example: https://stackoverflow.com/questions/48491688/how-to-define-parameters-with-square-brackets-in-openapi-swagger

**Comment by sm-Fifteen** [comment]
See also #283 for broader support of encoding styles.

This is the sort of issue where it's not entirely clear if the responsibility of supporting this should be on FastAPI or Pydantic. On one hand, it has to do with value parsing and encoding, which is something FastAPI tends to fully delegate to Pydantic, but on the other hand it's not part of JSON Schema, it's [an OpenAPI extension](https://oai.github.io/Documentation/specification-parameters.html#parameter-serialization-control) that [only applies to route parameter objects](https://github.com/OAI/OpenAPI-Specification/blob/main/schemas/v3.1/schema.yaml#L312-L413), and most of these encoding styles don't make sense in a JSON context (which is what pydantic is generally meant for).

I would like to see these supported as well.

**Comment by ghandic** [comment]
I'm keen to pursue this issue as I think it is a fairly common use-case (for example what to sort by or filter by)

See links on the different supported options below:

- [OpenAPI](https://spec.openapis.org/oas/latest.html#style-examples)
- [Swagger](https://swagger.io/docs/specification/serialization#query)

The current solution is to create your own parser if you want to support it. However, the only way I have found is to use a regex pattern to display on the OpenAPI spec. See below link for an example of a rough implementation of custom parsing for sortby and filterby

https://gist.github.com/ghandic/21c27470f6797dd856208a2c68f3e43a

Ideally this could look more as follows


```python
from fastapi import QueryMixIn
from pydantic import BaseModel

...

class FilterBy(BaseModel, QueryMixIn):
    name: str
    q: str

    class Config:
        query_style = "deepObject"


class SortBy(BaseModel, QueryMixIn):
    name: str
    by: str

    class Config:
        query_style = "deepObject"

@router.get("/search")
async def search(
    sort: Optional[SortBy] = Query(None, ),
    filter: List[FilterBy] = = Query(None, ),
    limit: int = Query(10, ge=1, le=50),
    page: int = Query(1, ge=1),
) -> List[Record]:
    ...
```

Then the user could standardize their Query style across their API using inheritance


```python
from fastapi import QueryMixIn
from pydantic import BaseModel

class QueryModel(BaseModel, QueryMixIn):
    class Config:
        query_style = "deepObject"
```


```python
from custom import QueryModel

...

class FilterBy(QueryModel):
    name: str
    q: str

class SortBy(QueryModel):
    name: str
    by: str

@router.get("/search")
async def search(
    sort: Optional[SortBy] = Query(None, ),
    filter: List[FilterBy] = = Query(None, ),
    limit: int = Query(10, ge=1, le=50),
    page: int = Query(1, ge=1),
) -> List[Record]:
    ...
```


The tough problem I see would be validation, but since we're using pydantic could the user not use their validators on their custom QueryModel inherited class

Thoughts?

**Comment by ghandic** [comment]
Please see example implementation here: https://github.com/ghandic/FastAPI-deepObject

**Comment by ghandic** [comment]
Ideally, I'd like to not define the types twice, would need to understand more of how the routes get registered for that implementation.

**Comment by ghandic** [comment]
Here is a example of what the OpenAPI spec should look like:

```yaml
openapi: 3.0.2
info:
  title: FastAPI
  version: 0.1.0
paths:
  /:
    post:
      summary: Demo
      operationId: demo__post
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Sort'
        required: true
      responses:
        '200':
          description: Successful Response
          content:
            application/json:
              schema: {}
        '422':
          description: Validation Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/HTTPValidationError'
    get:
      summary: Demo
      operationId: demo__get
      parameters:
      - in: query
        name: sort
        schema:
          $ref: '#/components/schemas/Sort'
        style: deepObject.   # ENABLE THIS
        explode: true   # ENABLE THIS
      responses:
        '200':
          description: Successful Response
          content:
            application/json:
              schema: {}
        '422':
          description: Validation Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/HTTPValidationError'
components:
  schemas:
    Direction:
      title: Direction
      enum:
        - asc
        - desc
      type: string
      description: An enumeration.
    HTTPValidationError:
      title: HTTPValidationError
      type: object
      properties:
        detail:
          title: Detail
          type: array
          items:
            $ref: '#/components/schemas/ValidationError'
    Sort:
      title: Sort
      required:
        - by
        - direction
      type: object
      properties:
        by:
          title: By
          pattern: ^[a-z]+$
          type: string
          example: name
        direction:
          $ref: '#/components/schemas/Direction'
    ValidationError:
      title: ValidationError
      required:
        - loc
        - msg
        - type
      type: object
      properties:
        loc:
          title: Location
          type: array
          items:
            type: string
        msg:
          title: Message
          type: string
        type:
          title: Error Type
          type: string

```

For some demo app (proposed solution to be worked on)

```python
from enum import auto

from fastapi import FastAPI, Query
from pydantic import BaseModel, Field
from fastapi_utils.enums import StrEnum

app = FastAPI()


class Direction(StrEnum):
    asc = auto()
    desc = auto()


class Sort(BaseModel):
    by: str = Field(..., example="name", regex="^[a-z]+$")
    direction: Direction


@app.post("/")
def demo(sort: Sort):
    return {}


# NOTE: Currently says:
# AssertionError: Param: sort can only be a request body, using Body(...)
# .... /lib/python3.8/site-packages/fastapi/dependencies/utils.py", line 331, in get_dependant
# This is where we should allow using deepObject
@app.get("/")
def demo(sort: Sort = Query(...)):
    return {}

```

**Comment by ghandic** [comment]
@tiangolo would you be supportive of this if the effort was put in to build the PR? It would take a good amount of effort and even more testing.

**Comment by adriangb** [comment]
For what it's worth I was able to support the full range of options in Xpresso, maybe that can serve as inspiration for FastAPI: https://xpresso-api.dev/0.42.3/tutorial/query_params/#customizing-deserialization

**Comment by commonism** [comment]
> How about just `Query(style="deepObject")`?

Rejected in #9768

**Comment by adriangb** [comment]
> > How about just `Query(style="deepObject")`?
> 
> Rejected in #9768

I don't think Sebastián closed the issue actually, I'll reply there

**Comment by frankie567** [comment]
I agree that supporting a `Query(style="deepObject")` would be really helpful, even with just the OpenAPI schema output, and not the automatic parsing.

**Comment by wdhwg001** [comment]
Perhaps this would be useful:

https://github.com/techouse/qs_codec

It's a qs compatible query parsing library, which can be used to parse axios-styled query string.

**Comment by WilliamDEdwards** [comment]
As a FastAPI user, I'd like to point out the issues I ran into while trying to implement `deepObject`-style query parameters.

In my case, I want to use them for filtering, e.g.:

    ?filter[id]=1&filter[name]=foo

... instead of a construct like:

    ?filter=id:1&filter=name:foo

I'm running into two main issues.

---

# Documentation

The 'square bracket' notation is called `deepObject`, see: https://swagger.io/docs/specification/v3_0/serialization/#query-parameters

To document that notation, the following must be added to the `parameter` stanza:

```
style: deepObject
explode: true
```

Example: https://stackoverflow.com/a/48498055

However, we can't alter that stanza with FastAPI:

* `json_schema_extra` alters the `schema` stanza, not `parameter` (parent)
* `openapi_extra` on the controller decorator (like `{'parameters': [{'name': 'filter', 'style': 'deepObject'}]}`) because a _new_ `filter` parameter is added, rather than deep-merged

---

# Alias

In the OpenAPI spec, parameters should have names like `id`. The `style: deepObject` definition is supposed to make clients understand that the parent (`filter`) and name should be merged into the 'square bracket' notation.

However, there is no way to do that:

| Pydantic `Field` option | Changed in OpenAPI spec | Expected input property changed |
| -- | -- | -- |
| `validation_alias` | :white_check_mark: | :white_check_mark: |
| `serialization_alias` | :x: | :x: |
| `alias` | :x: | :x: |

In other words, if we want a parameter in the OpenAPI spec to have the name:

    foo

... but expect the following on input:

    filter[foo]

... there is no way.

**Comment by commonism** [comment]
PR for deepObject is #9867 
Fresh rebased, in case you want to give it a try.

CI says does not work any longer, get me a second.
All - but 3.14 - green.

**Comment by tiangolo** [comment]
Thanks for the conversation here everyone! 🙌 

This is something I've had in the backlog of ideas for a while. But still, there are other things that have had a higher priority and some things I need to solve first.

I'm gonna keep tracking this internally, but I'll close this specific issue for now, for cleanup, and because I can't priorize this one now.

---

Having issues open for specific things is becoming problematic as they attract AI slop PRs, so I'm closing this for cleanup. ☕
